### PR TITLE
fix(sync): headers stage progress

### PR DESCRIPTION
**Pull request by rkrasiuk**
The previous behavior of the `HeaderStage` in case of an error would be propagating it to the pipeline. The stage might have already committed headers before encountering the error. Upon the next invocation of `HeaderStage::execute`, the stage would attempt to write headers from the consensus tip (without closing the db gap) resulting in overwrite and further failure.

This PR changes how the `HeaderStage` looks up the range of headers to download. The algorithm is the following:
- look up the `stage_progress` in the database
- look at the next item
  * if there is no next header, there is no gap in the database and we should download `[local head, consensus tip]` range
  * if the next header exists and isn't sequential, that header's parent hash will be the tip to download from
  * if the next header is congruent, the stage progress is invalid

### CODE REVIEW

File: crates/stages/src/stages/headers.rs [hunk 1]
```diff
@@ -74,26 +74,18 @@ impl<DB: Database, D: HeaderDownloader, C: Consensus, H: HeadersClient, S: Statu
...
-        let head = SealedHeader::new(last_header, last_hash);
-
-        let forkchoice = self.next_fork_choice_state(&head.hash()).await;
-        if let Some(number) = db.get::<tables::HeaderNumbers>(forkchoice.head_block_hash)? {
-            if number < head.number {
-                // Nothing to do here
-                warn!("Consensus reported old head {}", forkchoice.head_block_hash);
-                return Ok(ExecOutput { stage_progress, done: true, reached_tip: true })
-            }
-        }
```
Comment (gakonst) [inline crates/stages/src/stages/headers.rs:92]: Why is this removed? @onbjerg said that this may require a rewind if it actually happened? Might be wrong.

File: crates/stages/src/stages/headers.rs [hunk 2]
```diff
@@ -130,7 +122,14 @@ impl<DB: Database, D: HeaderDownloader, C: Consensus, H: HeadersClient, S: Statu
         // Write total difficulty values after all headers have been inserted
         self.write_td::<DB>(db, &head)?;
 
-        Ok(ExecOutput { stage_progress: current_progress, reached_tip: true, done: true })
+        let stage_progress = current_progress.max(
+            db.cursor::<tables::CanonicalHeaders>()?
+                .last()?
+                .map(|(num, _)| num)
+                .unwrap_or_default(),
+        );
```
Comment (gakonst) [inline crates/stages/src/stages/headers.rs:130]: nice. @onbjerg pls confirm this is correct. is this supposed to be `.last()?`

File: crates/stages/src/test_utils/macros.rs [hunk 3]
```diff
@@ -142,4 +111,42 @@ macro_rules! stage_test_suite {
     };
 }
 
+macro_rules! stage_test_suite_ext {
```
Comment (onbjerg) [inline crates/stages/src/test_utils/macros.rs:114]: docs on this one? how it differs from e.g. `stage_test_suite!`

File: crates/stages/src/error.rs [hunk 4]
```diff
@@ -36,6 +36,9 @@ pub enum StageError {
     /// rely on external downloaders
     #[error("Invalid download response: {0}")]
     Download(String),
+    /// Invalid checkpoint passed to the stage
+    #[error("Invalid checkpoint: {0}")]
+    Checkpoint(u64),
```
Comment (onbjerg) [inline crates/stages/src/error.rs:41]: This is technically fatal, right? Since the pipeline will just load up the same checkpoint again

File: crates/stages/src/stages/headers.rs [hunk 5]
```diff
@@ -167,6 +166,26 @@ impl<D: HeaderDownloader, C: Consensus, H: HeadersClient, S: StatusUpdater>
...
+
+        // Lookup the head hash and reposition the cursor to it
+        let (_, head) = cursor
+            .seek_exact(stage_progress)?
+            .ok_or(DatabaseIntegrityError::CanonicalHash { number: stage_progress })?;
+
+        let tip = match cursor.next()? {
+            Some((number, hash)) if stage_progress + 1 != number => hash,
+            _ => self.next_fork_choice_state(&head).await.head_block_hash,
+        };
```
Comment (gakonst) [inline crates/stages/src/stages/headers.rs:187]: Is this explanation of the code right?

```suggestion
        // If the next element found in the cursor is not the "expected" next block per our current progress, 
        // then there was a conflict and we should start from there. Else, it should use whatever the forkchoice
        // state decides.
        let tip = match cursor.next()? {
            Some((number, hash)) if stage_progress + 1 != number => hash,
            _ => self.next_fork_choice_state(&head).await.head_block_hash,
        };
```

File: crates/stages/src/test_utils/macros.rs [hunk 6]
```diff
@@ -142,4 +111,42 @@ macro_rules! stage_test_suite {
...
+
+            // Run `after_execution` hook
+            runner.after_execution(seed).await.expect("failed to run after execution hook");
+
+            // Assert the successful result
+            let result = rx.await.unwrap();
+            assert_matches::assert_matches!(
+                result,
+                Ok(ExecOutput { done, reached_tip, stage_progress })
+                    if done && reached_tip && stage_progress == stage_progress
```
Comment (gakonst) [inline crates/stages/src/test_utils/macros.rs:142]: What happens if the `if` block is `false`?

File: crates/stages/src/stages/headers.rs (hunk 2 above)
Comment (onbjerg) [reply crates/stages/src/stages/headers.rs:130]: yes but only because we are committing inside of the stage currently, so this code path will only be hit once the entire range is synced (and there are no gaps)

File: crates/stages/src/stages/headers.rs [hunk 7]
```diff
@@ -167,6 +166,29 @@ impl<D: HeaderDownloader, C: Consensus, H: HeadersClient, S: StatusUpdater>
...
+            .seek_exact(stage_progress)?
+            .ok_or(DatabaseIntegrityError::CanonicalHash { number: stage_progress })?;
+
+        let tip = match cursor.next()? {
+            Some((number, hash)) if stage_progress + 1 != number => hash,
+            None => self.next_fork_choice_state(&head).await.head_block_hash,
+            // if next header is present and there is no gap - checkpoint is incorrect
+            _ => return Err(StageError::Checkpoint(stage_progress)),
+        };
+
```
Comment (onbjerg) [inline crates/stages/src/stages/headers.rs:188]: trace here would be nice ("Syncing from x to y", "Filling gap between a and b")

File: crates/stages/src/stages/headers.rs (hunk 2 above)
Comment (gakonst) [reply crates/stages/src/stages/headers.rs:130]: let's add this exact note in a doc comment above this line

File: crates/stages/src/error.rs (hunk 4 above)
Comment (gakonst) [reply crates/stages/src/error.rs:41]: Checkpoint is a new word, let's call it StageProgress too?

**Comment by gakonst** [comment]
LGTM pending our nits - just rebased / fixed the conflict

File: crates/stages/src/test_utils/macros.rs (hunk 6 above)
Comment (onbjerg) [reply crates/stages/src/test_utils/macros.rs:142]: the assertion will fail: https://docs.rs/assert_matches/latest/assert_matches/

Comment (rkrasiuk) [reply crates/stages/src/test_utils/macros.rs:142]: the assertion is incorrect and it panics

File: crates/stages/src/stages/headers.rs (hunk 7 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/headers.rs]: added in a different place

File: crates/stages/src/test_utils/macros.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/stages/src/test_utils/macros.rs:116]: added

**Review by onbjerg — APPROVED** [review]
